!DRAFT! Templates

SidhraFernando-Plant opened the pull request:
# DRAFT
Still working on this

LvandoApps commented on frontend/src/hooks/useCreateBatch.ts:
```diff
@@ -0,0 +1,78 @@
+  selectStates(createData.states.map(state => state.id))
+  createData.comments.forEach(comment => {
+    // TODO: ensure position isn't out of window
+    comment.x += PASTE_POSITION_OFFSET
+    comment.y += PASTE_POSITION_OFFSET
```
Small nitpick, can replace this with the below:
`comment.x, comment.y = comment.x + PASTE_POSITION_OFFSET, comment.y + PASTE_POSITION_OFFSET`
See it a few times, nothing major just a fussy edit

SidhraFernando-Plant replied:
Good point, will change this

SidhraFernando-Plant commented on frontend/src/hooks/useActions.js:
```diff
@@ -563,4 +504,39 @@ const promptLoadFile = (parse, onData, errorMessage = 'Failed to parse file') =>
+// Takes in the IDs of states, comments, and transitions
+// Parameters also include  the current project and whether a template is being created
+// Outputs a CopyData to be copied or Template object to be created into a template
+const selectionToCopyTemplate = (stateIds, commentIds, transitionIds, project, isTemplate) => {
```
This method takes a selection and makes it into a CopyData or Template for the COPY or CREATE_TEMPLATE actions

SidhraFernando-Plant commented on frontend/src/hooks/useCreateBatch.ts:
```diff
@@ -0,0 +1,82 @@
+  const setStateInitial = useProjectStore(s => s.setStateInitial)
+  const commit = useProjectStore(s => s.commit)
+
+  const createBatch = () => {
```
I tried to make a hook for inserting + selecting a group of elements (states, transitions, comments) as this is used in the copy and template insert functions. I couldn't get it to work properly because I always had issues calling so I made it into a store method insertGroup in the useProjectStore. I would need some help/pairing to make this happen, but for now it works in the store although it doesn't seem like the best solution

SidhraFernando-Plant replied:
(to clarify, this isn't actually called anywhere but I have left it in for now in case)

ire4ever1190 replied:
If you don't need it anymore then feel free to delete 👍
We can recover it through git anyways if we need it later

ire4ever1190 commented on frontend/src/components/Sidepanel/Panels/Templates/Templates.js:
Shouldn't this be in typescript?

SidhraFernando-Plant replied:
Oh yep sorry I was meant to update it, and then forgot 😅

ire4ever1190 commented on frontend/src/hooks/useActions.js:
```diff
@@ -41,17 +40,30 @@ const useActions = (registerHotkeys = false) => {
+    CREATE_TEMPLATE: {
+      hotkey: { key: 'j', meta: true },
+      handler: () => {
+        if (selectedStatesIds.length === 0 && selectedCommentsIds.length === 0 && selectedTransitionsIds.length === 0) {
```
Might be better to move this into the disabled field

SidhraFernando-Plant replied:
Is there a way I can use this command for testing if it's disable? If not, I'm going to leave it as not disabled for now so that I can keep testing it. By the time this is merged to dev the UI will hopefully be merged so it shouldn't need to be disabled (although if it wasn't finished when merging I would make sure it's disabled)

ire4ever1190 replied:
The logic will mostly stay the same, just means the UI button will be greyed out if nothing is selected (So you should be able to keep testing it)
```ts
disabled: () => selectedStatesIds.length === 0 && selectedCommentsIds.length === 0 && selectedTransitionsIds.length === 0,
handler: () => {
  // Normal code here, no need to check if stuff is selected anymore
```

SidhraFernando-Plant replied:
Oh right, I thought you meant disable it entirely but that makes more sense

ire4ever1190 commented on frontend/src/hooks/useActions.js:
```diff
@@ -563,4 +504,39 @@ const promptLoadFile = (parse, onData, errorMessage = 'Failed to parse file') =>
+// Parameters also include  the current project and whether a template is being created
+// Outputs a CopyData to be copied or Template object to be created into a template
+const selectionToCopyTemplate = (stateIds, commentIds, transitionIds, project, isTemplate) => {
+  const selectedStates = stateIds.map((stateId) => {
+    return project.states.find((state) => {
+      return state.id === stateId
+    })
+  })
```
Imo it'll be cleaner to use `filter` instead of `map` and `find`
```suggestion
  const selectedStates = project.states.filter(state => stateIds.includes(state))
```

Could also move it into a simple helper like (This is in typescript so its clearer, I understand the file is still JS)
```suggestion
  const filterFrom = <T>(allowed: T[], from: T[]): T[] => from.filter(x => allowed.includes(x))
  const selectedComments = filterFrom(commentIds, project.comments)
```
Either way is fine, but make sure to update the comments and transitions filtering also

ire4ever1190 commented on frontend/src/hooks/useTemplateInsert.ts:
```diff
@@ -0,0 +1,66 @@
+  })
+}
+
+const snapPosition = ({ x, y }) =>
```
This exact function is in `useStateCreation` also. Might be best to move the function into a util file like `points.js` instead of repeating it in multiple places

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -24,11 +26,28 @@ import {
 export type StoredProject = Project & {_id: string}
 
+export enum InsertGroupResponseType {
+  // eslint-disable-next-line no-unused-vars
```
Weird you need to disable, but I think I know the problem and will fix myself in a seperate PR :+1:

SidhraFernando-Plant replied:
Oh yes I meant to single this out in a comment, linter wasn't liking enums 😢 Thank you, that would be great!

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -226,6 +244,87 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
+        // TODO: ensure position isn't out of window
+        comment.x += PASTE_POSITION_OFFSET
+        comment.y += PASTE_POSITION_OFFSET
+        const newId = i + 1 + Math.max(-1, ...get().project.comments.map((c: ProjectComment) => c.id))
```
Move this out into a function like
```ts
const nextIDFor = (ids: {id?: number}[]): number => i + 1 + Math.max(-1, ...ids.map(s => s.id))
```
And then use in the other places that creating a new ID is needed. 

Also not too sure but looks like that during these loops where a newID is needed the only difference is `i +`? So might be better to cache the highest found and then just increment/add `i` inside the loop

ire4ever1190 reviewed (changes requested):
Looking good besides those few things :rocket: 

Like the UI concept :ok_hand: will be cool once that is finished in the next PR

LvandoApps commented on frontend/src/components/Toolbar/Toolbar.js:
```diff
@@ -54,6 +54,14 @@ const tools = [
     animation: commentAnimation
+  },
+  {
+    label: 'Template tool',
+    hotkey: 'J',
+    description: 'Add templates to your automaton',
+    value: 'template',
+    icon: <Star />,
+    animation: commentAnimation
```
Would be good if it had its own unique animation when you hover over the tool as I found this really useful in understanding how to use the other tools. I'm not familiar with how it's able to do that though I've never seen JSON files used in that way (see the JSON files in the animation folder in the Toolbar directory), might be best to consult group 1 on how they did this if neither you nor Jake know (assuming they implemented this animation feature).

LvandoApps commented:
Looking good so far, excited to see how the UI turns out 🚀

SidhraFernando-Plant commented:
I believe I have addressed all feedback, except disabling the action but I am going to do that along with the UI. 

I am going to branch off of this branch for the UI (I have merged with dev so this branch is not stale), then once everything is finalised I'll merge the whole branch train. So pretty much I am re-requesting review to check that I have suitably addressed all feedback, NOT with the intention of merging this branch to dev.

ire4ever1190 commented on frontend/src/components/Sidepanel/Panels/Templates/templatesStyle.js:
`.js -> .ts`

SidhraFernando-Plant replied:
Yep will convert to TS. I'll put this PR back to a draft I reckon til the other one is finished

ire4ever1190 reviewed (approved):
Looks good, assuming you'll convert that `templatesStyle.js` during the UI part.

So is this PR staying open? Or close it and then open a new one once UI is finished?

ire4ever1190 commented:
Re-requested myself so I don't forget 😵‍💫

Will review after you merge in dev :rocket:

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -180,65 +159,17 @@ const useActions = (registerHotkeys = false) => {
-        selectTransitions(pasteData.transitions.map(transition => transition.id))
-        if (isNewProject && pasteData.initialStateId !== null && project.initialState === null) {
-          setStateInitial(pasteData.initialStateId)
+        // createBatch(pasteData, project)
```
```suggestion
```
Hook has been deleted iirc

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -180,65 +159,17 @@ const useActions = (registerHotkeys = false) => {
+        // createBatch(pasteData, project)
+        const insertResponse = insertGroup(pasteData)
+        // This will be better in TS with enum
+        if (insertResponse.type === 2) {
```
```suggestion
        if (insertResponse.type === InsertGroupResponseType.SUCCESS) {
```
Just so we aren't using a magical number

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -180,65 +159,17 @@ const useActions = (registerHotkeys = false) => {
+          selectStates(insertResponse.body.states.map(state => state.id))
+          selectTransitions(insertResponse.body.transitions.map(transition => transition.id))
+          commit()
+        } else if (insertResponse.type === 1) {
```
```suggestion
        } else if (insertResponse.type === InsertGroupResponseType.FAIL) {
```
Same

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -588,4 +522,23 @@ const promptLoadFile = (parse, onData, errorMessage = 'Failed to parse file') =>
+// Takes in the IDs of states, comments, and transitions
+// Parameters also include  the current project and whether a template is being created
+// Outputs a CopyData to be copied or Template object to be created into a template
+export const selectionToCopyTemplate = (stateIds: number[], commentIds: number[], transitionIds: number[], project): CopyData => {
```
```suggestion
export const selectionToCopyTemplate = (stateIds: number[], commentIds: number[], transitionIds: number[], project: StoredProject): CopyData => {
```
If you set the type here, then you won't need to specify the types later on when filtering

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -588,4 +522,23 @@ const promptLoadFile = (parse, onData, errorMessage = 'Failed to parse file') =>
+  const selectedComments: ProjectComment[] = project.comments.filter(comment => commentIds.includes(comment.id))
+  const selectedTransitions = project.transitions.filter(transition => transitionIds.includes(transition.id))
+  const isInitialSelected = stateIds.includes(project.initialState)
+  const returnObject: CopyData = {
+    states: selectedStates,
+    comments: selectedComments,
+    transitions: selectedTransitions,
+    projectSource: project._id,
+    projectType: project.projectType,
+    initialStateId: isInitialSelected ? project.initialState : null
+  }
+  return returnObject
```
```suggestion
  return {
    states: selectedStates,
    comments: selectedComments,
    transitions: selectedTransitions,
    projectSource: project._id,
    projectType: project.projectType,
    initialStateId: isInitialSelected ? project.initialState : null
  }
```
Very small nitpick 😄 I just usually like to return objects directly

SidhraFernando-Plant replied:
Nah that's all good it makes sense, I think this was left over from when there were some different options of stuff to assign to the return

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -224,6 +245,94 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
+      return { type: InsertGroupResponseType.FAIL, body: `You cannot insert elements from a ${createData.projectType} project into a ${get().project.projectType} project.` }
+    }
+    // Check that for transitions being inserted, to and from states are also inserted
+    let missingState = false
+    createData.transitions.forEach((transition) => {
+      let toFound = false
+      let fromFound = false
+      createData.states.forEach((state) => {
+        if (state.id === transition.from) {
+          fromFound = true
+        }
+        if (state.id === transition.to) {
+          toFound = true
+        }
+      })
+      // If a transition has to or from missing, flag to return error
+      if (!(toFound && fromFound)) {
+        missingState = true
+      }
+    })
+    if (missingState) {
+      return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
+    }
```
```suggestion
    const stateIDs = new Set(createData.states.map(it => it.id))
    for (const transition of createData.transitions) {
      const hasTo = stateIDs.has(transition.to)
      const hasFrom = stateIDs.has(transition.from)
      // If either `to` or `from` is missing then return an error
      if (!(hasTo && hasFrom)) {
        return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
      }
    }
```
Using sets is a bit cleaner imo, lemme know what you think though.
I tired to do the same logic so should work the same (if something is fishy plz mention). Main difference is that we return the error inside the loop so that it breaks early if we know something is wrong

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -224,6 +245,94 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
+      })
+      const nextTransitionId = nextIDFor(project.transitions)
+      createData.transitions.forEach((transition, i) => {
+        const newId = nextTransitionId + i
+        transition.id = newId
```
```suggestion
        transition.id = nextTransitionId + i
```
Same

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -224,6 +245,94 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
+      createData.comments.forEach((comment, i) => {
+        // TODO: ensure position isn't out of window
+        [comment.x, comment.y] = [comment.x + positionOffset, comment.y + positionOffset]
+        const newId = nextCommentId + i
+        comment.id = newId
```
```suggestion
        comment.id = nextCommentId + i
```
Same small nitpick. Good in places like this though since it means one less variable needs to be kept in mental context when reading code

ire4ever1190 commented on frontend/src/stores/useTemplateStore.ts:
```diff
@@ -0,0 +1,27 @@
+interface TemplateStore {
+  template: Template,
+  setTemplate: (template: Template) => void
+  // update: (template: Template) => void,
+  // setName: (newName: string) => void
```
```suggestion
  // TODO: Support updating/renaming
```
Just so we can find what needs to be done by searching for `// TODO`

ire4ever1190 commented on frontend/src/stores/useTemplateStore.ts:
```diff
@@ -0,0 +1,27 @@
+  setTemplate: template => {
+    set({ template })
+  }
+  // update: (template) => {
+
+  // },
+  // setName: (newName) => {
+
+  // }
```
```suggestion
```

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -249,14 +180,17 @@ const useActions = (registerHotkeys = false) => {
-          removeTransitions(selectedTransitionIDs)
-          removeComments(selectedCommentIDs)
+        // If a template is selected and nothing else is, delete the template
+        if (template !== null && selectedCommentsIds.length === 0 && selectedStatesIds.length === 0 && selectedTransitionsIds.length === 0) {
```
Might be good to move out the length checks into a variable
```ts
const nothingSelected = selectedCommentsIds.length === 0 && selectedStatesIds.length === 0 && selectedTransitionsIds.length === 0
```
And then the else if can just be `else if (!nothingSelected)`

ire4ever1190 commented on frontend/src/hooks/useActions.ts:
```diff
@@ -588,4 +522,23 @@ const promptLoadFile = (parse, onData, errorMessage = 'Failed to parse file') =>
+// Parameters also include  the current project and whether a template is being created
+// Outputs a CopyData to be copied or Template object to be created into a template
+export const selectionToCopyTemplate = (stateIds: number[], commentIds: number[], transitionIds: number[], project): CopyData => {
+  const selectedStates: AutomataState[] = project.states.filter(state => stateIds.includes(state.id))
+  const selectedComments: ProjectComment[] = project.comments.filter(comment => commentIds.includes(comment.id))
```
```suggestion
  const selectedStates = project.states.filter(state => stateIds.includes(state.id))
  const selectedComments = project.comments.filter(comment => commentIds.includes(comment.id))
```
No need after previous suggestion

ire4ever1190 commented on frontend/src/stores/useProjectStore.ts:
```diff
@@ -57,6 +75,10 @@ export const createNewProject = (projectType: ProjectType = DEFAULT_PROJECT_TYPE
   }
 })
 
+const nextIDFor = (elementArr: AutomataState[] | BaseAutomataTransition[] | ProjectComment[]): number => {
+  return 1 + Math.max(-1, ...elementArr.map((e: AutomataState | BaseAutomataTransition | ProjectComment) => e.id))
+}
```
We only need an `id` so we can just specify that (Also means it will work for any future items)
```suggestion
const nextIDFor = (elementArr: {id?: number}[]): number => {
  // We can't do elementArr.length + 1 since that might reuse an ID from a deleted item
  // Uses || -1 to handle any of the items being `undefined` (Shouldn't happen, but our type says it can)
  return 1 + (Math.max(-1, ...elementArr.map(e => e.id)) || -1)
}
```
I used `|| -1` since `Math.max(-1, undefined)` returns `NaN` (This shouldn't happen in practise though, except the type for `ProjectComment` needs to allow for undefined ID so this is more just defensive programming).

Also I can do this if you want, but `createState`, `createTransition`, and `createComment` should also be using `nextIDFor`

SidhraFernando-Plant replied:
I tried putting this function in, but it resulted in some unexpected behaviour: if I have an empty project, create a state (q0), and then copy paste it the next ID is not calculated correctly (it is pasted as q0 as well). 

So I was thinking, perhaps it could be left as is and this could be opened as a separate issue in the interest of prioritising getting the templates into dev? I'm sure it wouldn't be too hard to work out what the issue is and get this refactored version into the code, I am just keen to get this merged if possible. Perhaps calling the function in createState, createTransition, and createTransition could be tackled in that same issue? Let me know what you think

ire4ever1190 reviewed (changes requested):
Nice work, just a few comments to look over but otherwise looking really good 🚀

ire4ever1190 reviewed (approved):
Sorry about that. All looking good 👍 you can merge whenever
